### ISSUE TITLE
Normalization of RetinaNet classification loss correct?

### ISSUE THREAD

**Issue by JohannesBrx** [facebookresearch/detectron2#703]
In [retinanet.py](https://github.com/facebookresearch/detectron2/blob/master/detectron2/modeling/meta_arch/retinanet.py#L173-L187), the (focal) loss is normalized by the number of foreground elements `num_foreground`. However, the number of elements being considered in the loss function are the valid elements `valid_idxs`, i.e., foreground and background elements. So I would expect the last code line to be something like `max(1, valid_idxs.sum())`. This would also the behavior when using `torch.nn.functional.binary_cross_entropy_with_logits` with `reduction='mean'`.

Is the normalization with `num_foreground` intentionally chosen or just taken over from the regression loss, where it seems to be plausible?

```
	valid_idxs = gt_classes >= 0
	foreground_idxs = (gt_classes >= 0) & (gt_classes != self.num_classes)
	num_foreground = foreground_idxs.sum()
	...
	# logits loss
	loss_cls = sigmoid_focal_loss_jit(
		pred_class_logits[valid_idxs],
		gt_classes_target[valid_idxs],
		alpha=self.focal_loss_alpha,
		gamma=self.focal_loss_gamma,
		reduction="sum",
	) / max(1, num_foreground)
```

**Comment by ppwwyyxx** [comment]
This matches the description in the original paper.

**Comment by zjysteven** [comment]
@JohannesBrx Previously I had the same question as you. However, according to my experiments (based on some other implementations, not detectron), if you normalize the focal loss with `valid_idxs` which includes background anchors, the classification loss will be extremely small even from the beginning of the training as the number of background anchors is typically very large, which actually will harm the training.
